src/doc-utils.js

```diff
@@ -41,26 +51,92 @@ function isEmpty(n) {
 }
 
 function getFirstString(doc) {
-  return iterDoc(doc, (type, doc) => {
-    if (type === "string" && doc.trim().length !== 0) {
-      return doc;
+  let firstString = null;
+  traverseDoc(doc, doc => {
+    if (
+      typeof doc === "string" && doc.trim().length !== 0 && firstString === null
+    ) {
+      firstString = doc;
+    }
+  });
+  return firstString;
+}
+
+function isLineNext(doc) {
+  let flag = null;
+  traverseDoc(doc, doc => {
+    if(flag === null) {
+      if (typeof doc === "string") {
+        flag = false;
+      }
+      if (doc.type === "line") {
+        flag = true;
+      }
     }
   });
+  return !!flag;
 }
 
-function hasHardLine(doc) {
-  // TODO: If we hit a group, check if it's already marked as a
-  // multiline group because they should be marked bottom-up.
-  return !!iterDoc(doc, (type, doc) => {
-    switch (type) {
+function willBreak(doc) {
+  let willBreak = false;
+  traverseDoc(doc, doc => {
+    switch (doc.type) {
+      case "group":
+        if(doc.break) {
+          willBreak = true;
+        }
       case "line":
         if (doc.hard) {
-          return true;
+          willBreak = true;
         }
 
         break;
     }
   });
+  return willBreak;
+}
+
+function breakParentGroup(groupStack) {
+  if(groupStack.length > 0) {
+    const parentGroup = groupStack[groupStack.length - 1];
+    // Breaks are not propagated through conditional groups because
+    // the user is expected to manually handle what breaks.
+    if(!parentGroup.expandedStates) {
+      parentGroup.break = true;
+    }
+  }
+  return null;
+}
+
+function propagateBreaks(doc) {
+  const groupStack = [];
+  traverseDocUpDown(
+    doc,
+    doc => {
+      if (doc.type === "break-parent") {
+        breakParentGroup(groupStack);
+      }
+      if (doc.type === "group") {
+        groupStack.push(doc);
+      }
+    },
+    doc => {
+      if (doc.type === "group") {
+        const group = groupStack.pop();
+        if(group.break) {
+          breakParentGroup(groupStack);
```

Comment:
How do you know what the parents are? We need to keep a "stack" so that we know what the current group is, and be able to move to the "previous" one when moving out of the group.

`breakParentGroup` could iterate up and mark all of the groups at once, but we still need to pop groups off the stack. We might as well mark them incrementally as we pop them off the stack, because otherwise you'll be iterate a lot over groups that have already been marked. There's less work if you just mark the current group, and then when leaving a group propagate it up to the "new" current group.